packages/ddp-server/livedata_server.js

```diff
@@ -1085,53 +1085,57 @@ _.extend(Subscription.prototype, {
     var isCursor = function (c) {
       return c && c._publishCursor;
     };
-    if (isCursor(res)) {
-      try {
-        res._publishCursor(self);
-      } catch (e) {
-        self.error(e);
-        return;
-      }
-      // _publishCursor only returns after the initial added callbacks have run.
-      // mark subscription as ready.
-      self.ready();
-    } else if (_.isArray(res)) {
-      // check all the elements are cursors
-      if (! _.all(res, isCursor)) {
-        self.error(new Error("Publish function returned an array of non-Cursors"));
-        return;
-      }
-      // find duplicate collection names
-      // XXX we should support overlapping cursors, but that would require the
-      // merge box to allow overlap within a subscription
-      var collectionNames = {};
-      for (var i = 0; i < res.length; ++i) {
-        var collectionName = res[i]._getCollectionName();
-        if (_.has(collectionNames, collectionName)) {
-          self.error(new Error(
-            "Publish function returned multiple cursors for collection " +
-              collectionName));
+
+    //Supports a publish handler result from a normal or an async function
+    Promise.resolve(resMaybePromise).then(res => {
```

Comment:
Such a change will make **all** publications async, which I think is not desired. The solution would be to either use `Promise.await` (but that relies on fibers), or use an optionally-asynchronous `then` (like the one I made [here](https://github.com/vazco/uniforms/blob/master/packages/uniforms/src/ValidatedForm.tsx#L185-L194), but without the `callIfAsync` option).

The type signature of such a `then` function is:
```ts
function then<T, U>(value: Promise<T>, fn: (value: T) => U): Promise<U>;
function then<T, U>(value: T, fn: (value: T) => U): U;
```

Reply:
@radekmie I strongly disagree. I tested the behaviour with both async and normal publish functions to confirm that the current behaviour of non-async publish function was retained.

Reply:
@radekmie this is one of my tests. On the server side I created these test publications **testa** and **testb**:

```
Meteor.publish('testa', async function() {
        console.log("Entered testa");
        Meteor._sleepForMs(5000);
        console.log("subscribing to testa");
        this.added('messages', Random.id(), { message: 'hello from test a' });
        this.ready();
});

Meteor.publish('testb', async function() {
        console.log("Entered testb");
        Meteor._sleepForMs(5000);
        console.log("subscribing to testb");
        this.added('messages', Random.id(), { message: 'hello from test b' });
        this.ready();
});
```

In a client side browser console, I ran both of these lines in one go:

```
a = Meteor.subscribe('testa');
b = Meteor.subscribe('testb');
```

When the publish functions are declared async, the server-side console prints:

```
I20210524-17:54:35.542(10)? Entered testa
I20210524-17:54:35.544(10)? Entered testb
I20210524-17:54:40.542(10)? subscribing to testa
I20210524-17:54:40.558(10)? subscribing to testb
```

This is the expected unblocked behaviour.

When the publish functions are not declared async, the server-side console prints:

```
I20210524-17:56:10.975(10)? Entered testa
I20210524-17:56:15.977(10)? subscribing to testa
I20210524-17:56:15.983(10)? Entered testb
I20210524-17:56:20.985(10)? subscribing to testb
```

This is the expected default blocked behaviour.

I agree that there must be some other subtle difference as something must be causing Travis CI to fail.